### ISSUE TITLE
Dictionary converts lists to tuples

### ISSUE THREAD

**Issue by DougAJ4** [xlwings/xlwings#741]
I am using the code shown below to call a python function from an Excel UDF.  

"func" may either be a named function or a text string that can be evaluated.
"kwargs" is a dictionary created from a three column range of keys (argument names), the associated values, and the number of dimensions for arrays .  If the value is a text string representing a range it is converted in VBA to a variant array, with 1 or 2 dimensions, depending on the value in column 3, so in VBA we end up with a 2 column variant array, with the second column containing either values or variant arrays.

When this array is converted to a Python dictionary, with @xw.arg('kwargs', dict, transpose=True), it seems that the arrays are converted to tuples, rather than lists.  This causes a problem with any function that writes over array arguments.

Is there any way to pass the arrays as lists (or lists of lists) within the dictionary, to be consistent with arrays passed directly?
```python

@xw.func
@xw.arg('kwargs', dict, transpose=True)   
def CallFuncD(func, modname, kwargs):
    try:
        if modname == None:
            func  = globals()[func]
        else:            
            modname = globals()[modname]
            func  = getattr(modname, func)
    except:
        try:
            func = eval(func)
        except:
            return "Invalid function"
    res = func(**kwargs)
    return res
```

**Comment by fzumstein** [comment]
Hi Doug, it's kind of hard to follow what your exact issue is. If you can strip it down to a minimal example and attach a fully working sample that shows what you get and what you would expect that might help.

**Comment by DougAJ4** [comment]
See the attached file.  I want to be able to call functions from external libraries, without having to set up a specific interface function.  The example starting at row 120 on the CallFuncD sheet illustrates the problem with a function from the Alglib library (with Python interface).
The relevant functions are:
CallFuncD: Python function and auto-generated VBA
xl_CallFuncD: VBA to create an array of argument names and values, to transfer to Python as a dictionary.
rmatrixlu: Alglib Python function (LU decomposition of a matrix).

I have added a "Convert" option to xl_CallFuncD, so that if this is set to True, CallFuncD converts the tuples in the argument dictionary into lists (or lists of lists).  This is necessary because rmatrixlu writes over the array argument ('a'), so if it is left as a tuple I get the error:
ValueError: 'a' parameter can't be cast to real_matrix
With the added Convert option the function now works as intended.

Passing the dictionary array values as tuples is obviously not a huge problem in this case (especially as I will need to add an option to convert to Numpy arrays to work with Scipy/Numpy), but it does seem to me that it would make more sense to use the same object type for arrays in dictionaries as is used for arrays passed directly.

p.s.
The Alglib example won't work unless you have Alglib python installed, so I'll add a Scipy example, and attach to a separate message.

[xlAlglib.zip](https://github.com/ZoomerAnalytics/xlwings/files/1243908/xlAlglib.zip)

**Comment by DougAJ4** [comment]
I have now added an example using the lu function from the scipy.linalg module (imported as sla).

Rather strangely, this one works no matter whether "Convert" and/or "overwrite_a" are set to True or False.  I'm guessing that the function needs a numpy array, so if it receives an argument of tuple or list type it automatically converts it.

As said in the previous message, this is not an urgent issue, but passing dictionary array values as lists would seem more consistent.
[xlAlglib.zip](https://github.com/ZoomerAnalytics/xlwings/files/1244092/xlAlglib.zip)

**Comment by fzumstein** [comment]
My guess is that this was caused by tinkering with the VBA wrapper code, so closing this. feel free to reopen if this is still an issue.
